plugins/image-prioritizer/class-ip-img-tag-visitor.php

```diff
@@ -68,12 +66,26 @@ public function __invoke( OD_HTML_Tag_Walker $walker ): bool {
 			 * Note also that if this is the LCP element for _some_ of the viewport groups, it will still get
 			 * fetchpriority=high by means of the preload link (with a media query) that is added further below.
 			 */
-			$walker->set_attribute( 'data-od-removed-fetchpriority', $walker->get_attribute( 'fetchpriority' ) );
 			$walker->remove_attribute( 'fetchpriority' );
 		}
 
-		// TODO: If the image is visible (intersectionRatio!=0) in any of the URL metrics, remove loading=lazy.
-		// TODO: Conversely, if an image is the LCP element for one breakpoint but not another, add loading=lazy. This won't hurt performance since the image is being preloaded.
+		// TODO: Also if the element isLCPCandidate it should never by lazy-loaded.
+		$element_max_intersection_ratio = $this->url_metrics_group_collection->get_element_max_intersection_ratio( $xpath );
+
+		// If the element was not found, we don't know if it was visible for not, so don't do anything.
+		if ( is_null( $element_max_intersection_ratio ) ) {
+			$walker->set_meta_attribute( 'unknown-tag', true ); // Mostly useful for debugging why an IMG isn't optimized.
+		} else {
+			// Otherwise, make sure visible elements omit the loading attribute, and hidden elements include loading=lazy.
+			$is_visible = $element_max_intersection_ratio > 0.0;
+			$loading    = (string) $walker->get_attribute( 'loading' );
+			if ( $is_visible && 'lazy' === $loading ) {
+				$walker->remove_attribute( 'loading' );
+			} elseif ( ! $is_visible && 'lazy' !== $loading ) {
+				$walker->set_attribute( 'loading', 'lazy' );
+			}
+		}
+		// TODO: If an image is the LCP element for one breakpoint but not another, add loading=lazy. This won't hurt performance since the image is being preloaded.
```

Comment:
That is, if the image is not the LCP element for all breakpoints _and_ the image is not in the initial viewport.